src/zenml/prompts/prompt.py

```diff
@@ -0,0 +1,164 @@
+
+        # System prompt
+        prompt = Prompt(
+            template="You are a helpful assistant.",
+            prompt_type="system"
+        )
+
+        # Versioned prompt for tracking changes
+        prompt = Prompt(
+            template="Answer: {question}",
+            version="2.0.0"  # Track iterations
+        )
+    """
+
+    model_config = {"protected_namespaces": ()}
+
+    # Core prompt content
+    template: str = Field(..., description="The prompt template string")
+
+    # Default variable values
+    variables: Dict[str, Any] = Field(
+        default_factory=dict,
+        description="Default variable values for template substitution",
+    )
+
+    # Basic classification
+    prompt_type: str = Field(
+        default="user",
+        description="Type of prompt: 'system', 'user', 'assistant'",
+    )
+
+    # Version for Git-like tracking
+    version: str = Field(
+        default="1.0.0",
+        description="Version of the prompt for tracking changes",
+    )
+
+    def format(self, **kwargs: Any) -> str:
+        """Format the prompt template with provided variables.
+
+        Args:
+            **kwargs: Variables to substitute in the template.
+                     These override any default variables.
+
+        Returns:
+            Formatted prompt string
+
+        Raises:
+            ValueError: If required variables are missing
+        """
+        # Merge default variables with provided kwargs
+        format_vars = {**self.variables, **kwargs}
+
+        try:
+            return self.template.format(**format_vars)
+        except KeyError as e:
+            missing_var = str(e).strip("'\"")
+            raise ValueError(
+                f"Missing required variable '{missing_var}' for prompt formatting. "
+                f"Available variables: {list(format_vars.keys())}"
+            )
+
+    def get_variable_names(self) -> List[str]:
+        """Extract variable names from the template.
+
+        Returns:
+            List of variable names found in the template
+        """
+        pattern = r"\{([^}]+)\}"
+        return list(set(re.findall(pattern, self.template)))
+
+    def validate_variables(self) -> bool:
+        """Check if all required variables are provided.
+
+        Returns:
+            True if all template variables have default values
+        """
+        template_vars = set(self.get_variable_names())
+        provided_vars = set(self.variables.keys())
+        return template_vars.issubset(provided_vars)
+
+    def get_missing_variables(self) -> List[str]:
+        """Get list of missing required variables.
+
+        Returns:
+            List of variable names that are required but not provided
+        """
+        template_vars = set(self.get_variable_names())
+        provided_vars = set(self.variables.keys())
+        return list(template_vars - provided_vars)
+
+    def __str__(self) -> str:
+        """String representation of the prompt.
```

Comment:
would you actually want to see the contents of the prompt within a stringified version of the prompt? seems like maybe the most important thing you'd want to return)